docs/architecture/adr-016-validator-consensus-key-rotation.md

```diff
@@ -0,0 +1,67 @@
+# ADR 016: Validator consensus key rotation
+
+## Changelog
+
+- 23-10-2019: initial draft
+
+## Context
+
+Validator consensus key rotation feature has been discussed and requested for a long time, for the sake of safer validator 
+key management policy.(https://github.com/tendermint/tendermint/issues/1136) So, we suggest one of the simplest form of
+validator consensus key rotation implementation mostly onto Cosmos-SDK.
+
+## Decision
+
+### Pseudo procedure for consensus key rotation
+
+- create new random consensus key.
+- create and broadcast a transaction(RotateValConsensusKey) that the new consensus key is now coupled with the validator operator with signature from validator wallet key.
+- old consensus key becomes unable to participate on consensus immediately after the update of key mapping state on-chain.
+- start validating with new consensus key.
+
+### Considerations
+
+- consensus key mapping information management strategy
+    - stores history of each key mapping changes in kvstore.
+    - gaiad can search corresponding consensus key paired with given validator operator for any arbitrary height in recent unbonding period.
+    - gaiad does not need any historical mapping information which is past more than unbonding period.
+- limits
+    - a validator cannot rotate its consensus key more than N time for any unbonding period, to prevent spams.
+    - a validator should contribute X atoms to community fund to rotate its consensus key, also to prevent spams.
+    - parameters can be decided by governance and stored in genesis file.
+- slash module
+    - slash module can search corresponding consensus key for any height so that it can decide which consensus key is supposed to be used for given height.
+- data pruning
+    - database does not need to keep the historical key mapping changes after unbonding period past.
+    - when pruning is on, a node prunes all historical key mapping changes past unbonding period.
+- further implementation
+    - introducing "validator control key" which has authority to sign transaction for given validator.
+    - validator control key can be replaced to another key by transaction(RotateValControlKey).
+
+### Special note on implementation
+
```

Comment:
This section is highly lacking details IMHO. The actual process (doesn't need any pseudo-code) needs to be explicit here. Outline the entire (can be relatively high-level) process of how a rotation would occur.

e.g.

```
1. The validator generates a new consensus keypair.
2. The validator generates and signs a `MsgRotateConsensusKey` tx with their operator key.
3. `ApplyAndReturnValidatorSetUpdates` ...
```

Reply:
Thanks for comments, How about this?

1. The validator generates a new consensus keypair.
2. The validator generates and signs a `MsgRotateConsPubKey` tx with their operator key and new ConsPubKey

        type MsgRotateConsPubKey struct {
        	ValidatorAddress  sdk.ValAddress  `json:"validator_address" yaml:"validator_address"`
        	NewPubKey         crypto.PubKey   `json:"new_pubkey" yaml:"new_pubkey"`
        }

3. `handleMsgRotateConsPubKey` gets `MsgRotateConsPubKey`, calls `RotateConsPubKey` with emits event
4. `RotateConsPubKey` 
    - checks if `NewPubKey` is not duplicated on `ValidatorsByConsAddr`
    - overwrites `NewPubKey` in `validator.ConsPubKey`
    - deletes old `ValidatorByConsAddr`
    - `SetValidatorByConsAddr` for `NewPubKey`
    - Add `ConsPubKeyRotationHistory` for tracking rotation

            type ConsPubKeyRotationHistory struct {
            	OperatorAddress         sdk.ValAddress `json:"operator_address" yaml:"operator_address"`
            	OldConsPubKey           crypto.PubKey  `json:"old_consensus_pubkey" yaml:"old_consensus_pubkey"`
            	NewConsPubKey           crypto.PubKey  `json:"new_consensus_pubkey" yaml:"new_consensus_pubkey"`
            	RotatedHeight           int64  `json:"rotated_height" yaml:"rotated_height"`
            }

5. `ApplyAndReturnValidatorSetUpdates` checks if there is `ConsPubKeyRotationHistory` with `ConsPubKeyRotationHistory.RotatedHeight == ctx.BlockHeight()` and if so, generates 2 `ValidatorUpdate` , one for a remove validator and one for create new validator 

        abci.ValidatorUpdate{
            PubKey: tmtypes.TM2PB.PubKey(OldConsPubKey),
            Power:  0,
        }
        
        abci.ValidatorUpdate{
            PubKey: tmtypes.TM2PB.PubKey(NewConsPubKey),
            Power:  v.ConsensusPower(),
        }

6. at `previousVotes` Iteration logic of `AllocateTokens`,  `previousVote` using `OldConsPubKey` match up with `ConsPubKeyRotationHistory`, and replace validator for token allocation
7. Migrate `ValidatorSigningInfo` and `ValidatorMissedBlockBitArray` from `OldConsPubKey` to `NewConsPubKey`

Reply:
Looks reasonable. Please make the changes 👍